service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/ConsentMapper.java

```diff
@@ -0,0 +1,190 @@
+import uk.nhs.connect.iucds.cda.ucr.POCDMT000002UK01Consent;
+import uk.nhs.connect.iucds.cda.ucr.POCDMT000002UK01Entry;
+import uk.nhs.connect.iucds.cda.ucr.POCDMT000002UK01Observation;
+import uk.nhs.connect.iucds.cda.ucr.POCDMT000002UK01Section;
+import uk.nhs.connect.iucds.cda.ucr.POCDMT000002UK01StructuredBody;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Date;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import static org.hl7.fhir.dstu3.model.IdType.newRandomUuid;
+
+@Component
+@AllArgsConstructor
+public class ConsentMapper {
+    private static final String OPT_OUT_URI = "http://hl7.org/fhir/ConsentPolicy/opt-out";
+    private static final String FHIR_SNOMED = "http://snomed.info/sct";
+    private static final String ITK_SNOMED = "2.16.840.1.113883.2.1.3.2.4.15";
+    private static final String NPFIT_CDA_CONTENT = "2.16.840.1.113883.2.1.3.2.4.18.16";
+    private static final String SYSTEM_CODE = "887031000000108";
+    private static final String PERMISSION_TO_VIEW = "COCD_TP146050GB01#PermissionToView";
+
+    public Consent mapConsent(POCDMT000002UK01ClinicalDocument1 clinicalDocument, Encounter encounter) {
+        Consent consent = new Consent();
+        consent.setIdElement(newRandomUuid());
+
+        Patient patient = (Patient) encounter.getSubjectTarget();
+
+        Identifier docIdentifier = new Identifier();
+        docIdentifier.setUse(Identifier.IdentifierUse.USUAL);
+        docIdentifier.setValue(clinicalDocument.getSetId().getRoot());
+
+        Date now = new Date();
+        consent.setLanguage(encounter.getLanguage());
+        consent.setStatus(Consent.ConsentState.ACTIVE)
+                .setPeriod(encounter.getPeriod())
+                .setPatientTarget(patient)
+                .setPatient(encounter.getSubject())
+                .addConsentingParty(encounter.getSubject())
+                .addOrganization(encounter.getServiceProvider())
+                .setData(List.of(new ConsentDataComponent()
+                        .setMeaning(ConsentDataMeaning.RELATED)
+                        .setReference(new Reference(encounter))))
+                .setIdentifier(docIdentifier)
+                .setDateTime(now)
+                .setPolicyRule(OPT_OUT_URI);
+
+        extractAuthCodesFromDoc(consent, clinicalDocument);
+        POCDMT000002UK01StructuredBody structuredBody = getStructuredBody(clinicalDocument);
+        if (structuredBody != null) {
+            extractDataPeriodFromDoc(consent, structuredBody);
+            extractConsentSource(consent, structuredBody);
+            extractTextBody(consent, structuredBody);
+        }
+
+        return consent;
+    }
+
+    private void extractAuthCodesFromDoc(Consent consent, POCDMT000002UK01ClinicalDocument1 clinicalDocument) {
+        if (clinicalDocument.sizeOfAuthorizationArray() > 0) {
```

Comment:
Just a hint: you could rewrite this method using functional style of programming and though not need to check the size of the array. Like (pseudocode):
Arrays.stream(clinicalDocument.getAuthorizationArray()).map(auth -> auth.getConsent).filter(consent -> consent.isSetCode).map(consent -> consent.getCode()).consume(this::getCodingFromCE)